### PR TITLE
base: Allow faction-based docking rights (12.10.20)

### PR DESCRIPTION
**Pull request by dd84ai**
feat(baseplugin): adding new docking permissions based on IFF.

adding new commands /base addfac, /base rmfac, /base lstfac, /base myfac
they make possible to add as friendly or hostile a whole faction by their IFF
this changes their docking permissions and friendly/hostile attitude to base

### CODE REVIEW

File: Plugins/Public/base_plugin/PlayerBase.cpp [hunk 1]
```diff
@@ -283,6 +283,13 @@ void PlayerBase::Load()
 						ini_get_wstring(ini, tag);
 						perma_hostile_tags.push_back(tag);
 					}
+					else if (ini.is_value("faction_ally_tag"))
+					{
+					wstring tag;
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerBase.cpp:288]: broken indentation

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 2]
```diff
@@ -451,6 +451,119 @@ namespace PlayerCommands
...
+			const wchar_t* widecstr = widestr.c_str();
+			PrintUserCmdText(client, L"%d, %s", *it, widecstr);
+		}
+		PrintUserCmdText(client, L"OK");
+	}
+	void BaseViewMyFac(uint client, const wstring &cmd)
+	{
+		uint aff = GetAffliationFromClient(client);
+
+		wstring theaffiliation = (HkGetWStringFromIDS(Reputation::get_name(aff))).c_str();
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:559]: extra brackets

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 3]
```diff
@@ -451,6 +451,119 @@ namespace PlayerCommands
...
+
+		std::wstring widestr = std::wstring(theaffiliation.begin(), theaffiliation.end());
+		const wchar_t* widecstr = widestr.c_str();
+
+		PrintUserCmdText(client, L"Ship aff: %d, %s", aff, widecstr);
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:564]:
```suggestion
		PrintUserCmdText(client, L"Ship affiliation: %d, %s", aff, widecstr);

```

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 4]
```diff
@@ -451,6 +451,119 @@ namespace PlayerCommands
...
+		PrintUserCmdText(client, L"OK");
+	}
+	void BaseViewMyFac(uint client, const wstring &cmd)
+	{
+		uint aff = GetAffliationFromClient(client);
+
+		wstring theaffiliation = (HkGetWStringFromIDS(Reputation::get_name(aff))).c_str();
+
+		std::wstring widestr = std::wstring(theaffiliation.begin(), theaffiliation.end());
+		const wchar_t* widecstr = widestr.c_str();
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:562]: just do this in the PrintUserCmdText call

File: Plugins/Public/base_plugin/PlayerBase.cpp [hunk 5]
```diff
@@ -536,14 +547,21 @@ float PlayerBase::GetAttitudeTowardsClient(uint client)
...
 
 	// Make base hostile if player is on the hostile list.
 	if (hostile_tags.find(charname) != hostile_tags.end())
 	{
 		return -1.0;
 	}
 
+	uint playeraff = GetAffliationFromClient(client);
+	// Make base friendly if player is on the friendly faction list.
+	if ((ally_factions).find(playeraff) != (ally_factions).end())
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerBase.cpp:560]: extra brackets

File: Plugins/Public/base_plugin/PlayerBase.cpp [hunk 6]
```diff
@@ -536,14 +547,21 @@ float PlayerBase::GetAttitudeTowardsClient(uint client)
 		{
 			return 1.0;
 		}
-	}
+	}
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerBase.cpp:550]: ???

File: Plugins/Public/base_plugin/Main.cpp [hunk 7]
```diff
@@ -1078,6 +1122,22 @@ static bool IsDockingAllowed(PlayerBase *base, uint client)
...
+	{
+		if (charname.find(*i) == 0)
+		{
+			return false;
+		}
+	}
+
+	//Allow dock if player is on the friendly faction list.
+	uint playeraff = GetAffliationFromClient(client);
+	if ((base->ally_factions).find(playeraff) != (base->ally_factions).end())
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/Main.cpp:1136]: extra brackets

File: Plugins/Public/base_plugin/Main.cpp [hunk 8]
```diff
@@ -976,6 +987,37 @@ bool UserCmd_Process(uint client, const wstring &args)
 		PlayerCommands::BaseLstAllyTag(client, args);
 		return true;
 	}
+	else if (args.find(L"/base addfac") == 0)
+	{
+		returncode = SKIPPLUGINS_NOFUNCTIONCALL;
+		PrintUserCmdText(client, L"Proceeding...");
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/Main.cpp:993]: ???

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 9]
```diff
@@ -451,6 +451,119 @@ namespace PlayerCommands
...
+		{
+			tag = std::stoi(GetParam(args, ' ', 2));
+		}
+		catch (exception)
+		{
+			PrintUserCmdText(client, L"ERR No tag");
+			return;
+		}
+
+		if ((base->ally_factions).find(tag) == (base->ally_factions).end())
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:529]: extra brackets

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 10]
```diff
@@ -451,6 +451,119 @@ namespace PlayerCommands
...
+
+	void BaseLstAllyFac(uint client, const wstring &cmd)
+	{
+		PlayerBase *base = GetPlayerBaseForClient(client);
+		if (CheckForBase(base, client)) return;
+
+		set<uint>::iterator it;
+		for (it = base->ally_factions.begin(); it != base->ally_factions.end(); ++it)
+		{
+			wstring theaffiliation = (HkGetWStringFromIDS(Reputation::get_name(*it))).c_str();
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:548]: extra brackets

**Review by dsyalex — CHANGES_REQUESTED** [review]
I'm extremely hesitant to allow any extra complexity around POB docking rules.

Please squash your commits.

File: Plugins/Public/base_plugin/PlayerBase.cpp (hunk 1 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerBase.cpp:288]: Done

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 2 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:559]: After searching for a bit, I think that redundant brackets (parenthesis) does nothing for the resulting code
https://www.quora.com/Extra-parentheses-is-have-any-effect-on-C-C++-program-or-a-time-of-compilation
Except for some unique situations.

Fixed anyway to -> wstring theaffiliation = HkGetWStringFromIDS(Reputation::get_name(aff)).c_str();

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 3 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:564]: Ok

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 4 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:562]: It works.

File: Plugins/Public/base_plugin/PlayerBase.cpp (hunk 5 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerBase.cpp:560]: fixed

File: Plugins/Public/base_plugin/PlayerBase.cpp (hunk 6 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerBase.cpp:550]: What? Changed by accident. Not sure how to remove it.

Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerBase.cpp:550]: The resulting change changes nothing anyway.

File: Plugins/Public/base_plugin/Main.cpp (hunk 7 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/Main.cpp:1136]: done

File: Plugins/Public/base_plugin/Main.cpp (hunk 8 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/Main.cpp:993]: It looks normal, I see nothing unusual. Perhaps do you mean not necessary console msg? Removed.

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 9 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:529]: done

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 10 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:548]: Done

**Comment by dd84ai** [comment]
> I'm extremely hesitant to allow any extra complexity around POB docking rules.
> 
> Please squash your commits.

This is why I added it as a new friendly list, better than adding new defensemods ;)
Tested more or less in local server, worked fine. Hmm I didn't make feature to show list of faction IDs in-game though, I'll try to do it perhaps, while I still have some time.
This feature was asked from me by other people, quite desired one. After some thinking I would not mind having it too ;)

**Comment by dd84ai** [comment]
Added desires feature, /aff all that shows affiliation IDs for all factions.
because CreateID is working as well, as I get the thing with converting between regular data and the ones which flhook uses usually.

**Comment by dd84ai** [comment]
Missed adding descriptions to help command, I ll add later

**Comment by dd84ai** [comment]
Added

**Comment by dd84ai** [comment]
https://discoverygc.com/forums/showthread.php?tid=183356
people opinion about it

**Comment by dd84ai** [comment]
Removed function for hash written on its own from your Python inspiration
because there is already MakeID in flhook which does the same.
and wrapped my trash into class. It looks slightly more compact in this way.

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 11]
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
...
+		{
+			if (AffList.size() == 0)
+			{
+				if (factions.size() == 0)
+					LoadListOfReps();
+				
+				for (map<string, uint>::iterator iter = factions.begin(); iter != factions.end(); iter++)
+				{
+					string factionnickname = iter->first;
+					uint ID = MakeId(factionnickname.c_str());
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:633]: Can we document what this does? Nothing else appears to use it. I assume this is like CreateID but for faction hashes?

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 12]
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
...
+					wstring factionname = GetFactioName(ID);;
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:634]: double semicolon, also function name has a typo - `GetFactionName`

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 13]
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
...
+				wstring theaffiliation = HkGetWStringFromIDS(Reputation::get_name(ID)).c_str();
+				if (theaffiliation == L"Object Unknown")
+				{
+					theaffiliation = L"No Affiliation";
+				}
+				return theaffiliation;
+			}
+			catch (exception e)
+			{
+				return L"No Affiliation";
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:619]: Unknown Reputation?

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 14]
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
...
+			}
+			ConPrint(L"base: AffList was loaded succesfully.\n");
+		}
+		public:
+		void Init()
+		{
+			LoadAffList();
+		}
+
+		void FindAndPrintOneAffiliation(uint client, uint smth)
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:646]: smth?

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 15]
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
...
+		{
+			std::list<Affiliations::AffCell>::iterator found;
+			found = std::find_if(AffList.begin(), AffList.end(), std::bind(IDComparision, std::placeholders::_1, smth));
+			if (found != AffList.end())
+				PrintUserCmdText(client, L"Aff ID: %u, %s, %s", found->id, (found->nickname).c_str(), (found->factionname).c_str());
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:651]: Aff? Let's not send "Aff" to players

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 16]
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
...
+			else
+				PrintUserCmdText(client, L"Aff ID: %u, Unknown, Unknown", smth);
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:653]: see above

File: Plugins/Public/base_plugin/PlayerCommands.cpp [hunk 17]
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
...
+		}
+		void PrintAll(uint client)
+		{
+			for (list<Affiliations::AffCell>::iterator iter = AffList.begin(); iter != AffList.end(); iter++)
+			{
+				PrintUserCmdText(client, L"Aff ID: %d, %s, %s", iter->id, (iter->nickname).c_str(), (iter->factionname).c_str());
```
Comment (dsyalex) [inline Plugins/Public/base_plugin/PlayerCommands.cpp:659]: see above

**Review by dsyalex — CHANGES_REQUESTED** [review]
Please write a commit message

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 11 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:633]: createIDTable = None
def FLFactionHash(nickName):
        FLFACTIONHASH_POLYNOMIAL = 0x1021

        # Build the crc lookup table if it hasn't been created
        global createIDTable
        if createIDTable is None:
                createIDTable = []
                for i in range(0, 2**8):
                        x = i << 8
                        for bit in range(0, 8):
                                x <<= 1
                                if x & 0x10000:
                                        x = (x ^ FLFACTIONHASH_POLYNOMIAL) & 0xFFFF
                        createIDTable.append(x)

        # Calculate the hash.
        hash = 0xFFFF
        for c in bytearray(nickName.lower().encode('utf-8')):
                hash = (hash >> 8) ^ createIDTable[(hash & 0xFF) ^ c]
        return hash

if __name__ == "__main__":
        import sys
        for n in sys.argv[1:]:
                h = FLFactionHash(n)
                
print(n, h, '0x{:04x}'.format(h))
alex@alex-laptop:~/game-repository (Development)$ python3 SERVICE/alexscripts/createfactionid.py fc_freelancer
fc_freelancer 4169 0x1049

Do you remember you gave me this function? Later I rewrote it to C++..... and a long time later after that Dartes told me about an already existing function in Flhook which does the same ;) He got the info from Laz. Basically MakeID is this function above, but in built in FLhook, same results.

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 13 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:619]: Fixed

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 14 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:646]: let's call it 'AffiliationID' then. (Which was created by hash function)

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 12 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:634]: Got it.

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 11 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:633]: Provided comment to code
//MakeID function (in built in Flhook) is the same as mentioned here in C# to CreateFactionID https://github.com/DiscoveryGC/FLHook/blob/master/Plugins/Public/playercntl_plugin/setup_src/FLUtility.cs

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 15 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:651]: Lets call it IFF ID, people approve.

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 16 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:653]: see above

File: Plugins/Public/base_plugin/PlayerCommands.cpp (hunk 17 above)
Comment (dd84ai) [reply Plugins/Public/base_plugin/PlayerCommands.cpp:659]: see above

**Comment by dd84ai** [comment]
Resolved everything

**Review by dsyalex — CHANGES_REQUESTED** [review]
requires rebase

**Comment by dd84ai** [comment]
Performed tests:
1) /base myfac works well. it shows expected output
2) Attempt to use addfac, rmfac and lstfac failed because I am not logged as base admin
3) logged as admin and attempt to add incorrect faction. Denied.
4) added two correct factions among whom my current faction, turned defensemode as hostile 2, and attempted to redock. I was green and succeeded
5) Reloaded plugin to see if it will persist through reloading. Succeded.
6) Deleted correct faction from list, succeded
7) Deleted not correct faction from list, not succeded
8) Deleted all factions and tried to redock

**Review by dsyalex — APPROVED** [review]

**Pull request merged**